acados/ocp_nlp/ocp_nlp_sqp.c

```diff
@@ -810,14 +811,15 @@ static void linearize_update_qp_matrices(void *config_, ocp_nlp_dims *dims, ocp_
 #endif
     for (i = 0; i <= N; i++)
     {
-        // cost
-        config->cost[i]->update_qp_matrices(config->cost[i], dims->cost[i], nlp_in->cost[i],
-                                            opts->cost[i], mem->cost[i], work->cost[i]);
         // dynamics
         if (i < N)
             config->dynamics[i]->update_qp_matrices(config->dynamics[i], dims->dynamics[i],
                                                 nlp_in->dynamics[i], opts->dynamics[i],
                                                 mem->dynamics[i], work->dynamics[i]);
+        // cost
```

Comment:
Why did you move the cost update after the dynamics? For the way it is coded, it should be before: `cost` initializes the hessian of the QP, and dynamics makes an update to it in case of exact hessian

Reply:
The problem is that when you want to have algebraic variables in the cost, `dynamics` necessarily needs to be called *before* `cost`, due to the fact that the sensitivities dz/dux need to be computed (it should actually be the same with nonlinear least-squares costs). I get your point though, this way the exact hessian update would not be correct. I can think about how to fix this, do you have a solution already?

Reply:
I guess we need a zero-initialization of the Hessian ( H=0 ) before anything, and just update ( H += ... ) in cost/dynamics/constraints, so order does not matter, and you can freely swap dynamics before cost

Reply:
I have made an attempt to do this here: 7bd9153. Please have a look, I do not know how to test this thoroughly!

Reply:
It looks fine, but it should be applied also to `sqp_rti` and `cost_external`

Reply:
addressed here 7bd9153 - b30f85a. Again, I have no idea about how to test this properly, please have a look/tell me how to test it :)